### ISSUE TITLE
Upload Request Body not set when client is in Batch mode

### ISSUE THREAD

**Issue by caleblloyd** [google/google-api-php-client#786]
When the client is in Batch mode, the request is returned before upload parameters are set:

https://github.com/google/google-api-php-client/blob/master/src/Google/Service/Resource.php#L210

Google_Http_MediaFileUpload modifies the Request to prepare it correctly for upload.  The specific use case I am running into is uploading a file to Google Cloud Storage.  Google_Http_MediaFileUpload changes the upload URI  and sets the POST request body.

When I try to batch uploads, Google_Http_MediaFileUpload is never called on the request and the request gets sent to the wrong URI with no request body.

**Comment by bshaffer** [comment]
Hi @caleblloyd,
Try linking again to the line in question. The `master` branch was updated yesterday and the line number you pointed to above is no longer relevant.

Are you using any particular sample code for the GCS file upload? Can you post your code so I can help debug it?

**Comment by caleblloyd** [comment]
@bshaffer just recreated this with 2.0.0RC4 using Guzzle6.

The shouldDefer() conditional is returning the request before it is transformed for upload.  Looks to me like Lines 204-212 need to be moved to right before the function's return statement on line 232. https://github.com/google/google-api-php-client/blob/master/src/Google/Service/Resource.php#L204-L212

Here's the code:

    use GuzzleHttp\Psr7;
    use \Google_Http_MediaFileUpload;
    use \Google_Service_Storage_StorageObject;

    // These are custom helpers I've implemented
    /** @var \Google_client $googleClient */
    $googleClient = $google->getClient();
    /** @var \GuzzleHttp\Client $httpClient */
    $httpClient = $google->getHttpClient();
    /** @var \Google_Service_Storage $storage */
    $storage = $google->getStorage();

    $bucket = 'example';
    $stream = Psr7\stream_for("hello");

    // Params to pass a stream for the HTTP Body in $storage->objects->insert
    $params = [
        'data' => $stream,
        'mimeType' => 'text/plain',
    ];

    // Metadata object for new Google Cloud Storage object
    $obj = new Google_Service_Storage_StorageObject();
    $obj->contentType = "text/plain";

    // Serial Upload - Works
    $obj->name = "serial";
    $response = $storage->objects->insert($bucket, $obj, $params);
    echo "Serial uploaded to: ".$response->mediaLink.PHP_EOL;

    // Batch Upload - Fails
    $googleClient->setUseBatch(true);
    $obj->name = "batch";
    /** @var \GuzzleHttp\Psr7\Request $request */
    $request = $storage->objects->insert($bucket, $obj, $params);
    try {
        $response = json_decode($httpClient->send($request)->getBody(), true);
        echo "Batch uploaded to: ".$response['mediaLink'].PHP_EOL;
    } catch (\Exception $e){
        echo $e->getMessage().PHP_EOL;
    }

    // Batch Upload Hacked - Works
    $upload = new Google_Http_MediaFileUpload($googleClient, $request, $obj->contentType, $stream);
    $request = $upload->getRequest();
    $response = json_decode($httpClient->send($request)->getBody(), true);
    echo "Batch Hacked uploaded to: ".$response['mediaLink'].PHP_EOL;

Output is:

    Serial uploaded to: https://www.googleapis.com/download/storage/v1/b/example/o/serial?generation=123&alt=media
    Client error: `POST https://www.googleapis.com/storage/v1/b/example/o` resulted in a `400 Bad Request` response:
    {
        "error": {
        "errors": [
       {
           "domain": "global",
        "reason": "wrongUrlForUpload",
        "message": "Upload reque (truncated...)
    
    Batch Hacked uploaded to: https://www.googleapis.com/download/storage/v1/b/example/o/batch?generation=123&alt=media

When I move Lines 204-212 to right before line 232, everything works as expected:

    Serial uploaded to: https://www.googleapis.com/download/storage/v1/b/example/o/serial?generation=123&alt=media
    Batch uploaded to: https://www.googleapis.com/download/storage/v1/b/example/o/batch?generation=123&alt=media

**Comment by bshaffer** [comment]
I see. Good catch! I will make the change and write a test for it

**Comment by caleblloyd** [comment]
Thanks, @bshaffer!

**Comment by bshaffer** [comment]
@caleblloyd after reviewing #860 and #327, fixing the URL should not have allowed you to do batch uploads. Were you ever able to perform batch uploads with the above code?

**Comment by caleblloyd** [comment]
@bshaffer the underlying call here is using Guzzle's Batch, which really is a bunch of single upload requests to different endpoints executed in a pool.  It's not actually uploading multiple files to the same endpoint (that's what these other issues seem to be referring to)

**Comment by temuri416** [comment]
@caleblloyd Do you mind sharing the working version of batch upload?

**Comment by temuri416** [comment]
@caleblloyd @bshaffer 

Whoa. After moving those lines as described above code from #860 works PERFECT!

Now... Is that a bug or a feature? What's Google gonna do about it?

**Comment by bshaffer** [comment]
Ahh, I see. Yes, that's a completely different thing. Guzzle's "batch" executes the requests in parallel using promises.  Thank you for your response!

@temuri416 you can use the code found in the [Guzzle docs](http://docs.guzzlephp.org/en/5.3/clients.html#batching-requests). For example:

```php
// Guzzle 5
$client = new Google_Client();
$service = new Google_Service_Drive($client);
$file1 = new Google_Service_Drive_DriveFile([
        'name' => '1.jpg',
        'mimeType' => 'image/jpeg',
        'parents' => [$newFolderId],
]);
$file2 = new Google_Service_Drive_DriveFile([
        'name' => '2.jpg',
        'mimeType' => 'image/jpeg',
        'parents' => [$newFolderId],
]);

$req1 = $service->files->create($file1, [
        'data' => file_get_contents('img/1.jpg'),
        'mimeType' => 'image/jpeg',
        'uploadType' => 'media',
        'fields' => 'id',
]);
$req2 = $service->files->create($file2, [
        'data' => file_get_contents('img/2.jpg'),
        'mimeType' => 'image/jpeg',
        'uploadType' => 'media',
        'fields' => 'id',
]);

$requests = [$req1, $req2];

// Results is a GuzzleHttp\BatchResults object.
$results = Pool::batch($client->getHttpClient(), $requests);

// Can be accessed by index.
echo $results[0]->getStatusCode();

// Can be accessed by request.
echo $results->getResult($requests[0])->getStatusCode();
```

I imagine this is possible with Guzzle6 as well.

**Comment by caleblloyd** [comment]
I actually get the Guzzle HTTP requests from the v2 PHP API and pass them to \GuzzleHttp\Pool.  Looks like this may circumvent the v2 API's Batch functionality.

\GuzzleHttp\Pool executes the requests through libcurl in an event loop, so they do happen asynchronously.

Here is a code sample of how I perform multiple uploads to Google Cloud Storage.  Note, you need to be running 2.0.0rc5 at least for this to work properly due to https://github.com/google/google-api-php-client/pull/789

    /** @var \Google_Client $client */
    $client->setUseBatch(true);

    /** @var \GuzzleHttp\Psr7\Request[] $requests */
    $requests = [];

    foreach ($docs as $doc) {
        $obj = new \Google_Service_Storage_StorageObject();
        $obj->name = $doc['name'];
        $obj->contentType = "text/plain";
        $requests[] = $storage->objects->insert(
            'bucket-name',
            $obj
        );
    }

    $httpClient = $client->getHttpClient();
    /** @var \GuzzleHttp\Client $httpClient */
    $batch = new \GuzzleHttp\Pool::batch($httpClient, $requests);

    foreach ($batch as $i => $response) {
        if ($response instanceof \Exception) {
            // something happened
        } else {
            // all good
        }
    }

**Comment by temuri416** [comment]
Thanks, guys.

So, for argument's sake, if I tweak Resources.php as described above, would V3 API continue supporting it?

My code from #860 works great now.

**Comment by bshaffer** [comment]
@temuri416 A pretty critical rule of thumb is never tweak a vendored library. Please implement your changes in your application code, and do not attempt to hack at the core vendor libraries.

**Comment by temuri416** [comment]
I know, I know.

That's not the point. My question is how Google would look at this "bonus" feature?

I can always extend or write my own wrapper. But since this behaviour contradicts the docs - can I count on this or will it get "fixed" (i.e. removed)?

**Comment by caleblloyd** [comment]
@temuri416 when you mention "Whoa. After moving those lines as described above code from #860 works PERFECT!" 

That was a bug resolved Dec 30th 2015 and first tagged in v2.0.0RC5.  If you update to v2.0.0RC5, you should get the fix in your vendored library.

**Comment by temuri416** [comment]
@caleblloyd 

That's strange. Because I tried with master and v2.0.0RC5 and it does not work.

This however fixed it for me in Resources.php:

![image](https://cloud.githubusercontent.com/assets/3987203/13269318/20b1c7c2-da55-11e5-9ca3-aaedfa3f8913.png)

I moved the lines you've indicated after it returns with ->shouldDefer().

**Comment by caleblloyd** [comment]
@temuri416 the line numbers referenced in this issue have since changed.  You basically reversed the code back to what it was in v2.0.0RC4.  Are you sure that a valid file was uploaded at all when you change these lines back?

The problem that I was running into in RC4 is that the Google_Http_MediaFileUpload object mutates the URL of the uploaded file to the correct URL.  If it hits shouldDefer() without Google_Http_MediaFileUpload this first (as you have reverted your code to do), the URL to upload the file to never gets properly set.

**Comment by temuri416** [comment]
I see what you're saying and you're right.

I didn't bother checking contents of uploaded files - they're all 0 bytes.

What is the "right" URL that should be set?

**Comment by bshaffer** [comment]
@temuri416 OK! Ha glad to hear it. You never know.  As to your other question, this is not something I want added to `Resource.php`. Throwing an exception would be the better way to go.

The changes to the url being referred to are made [here](https://github.com/google/google-api-php-client/blob/v2.0.0-RC5/src/Google/Http/MediaFileUpload.php#L331)

**Comment by prateekkathal** [comment]
Please re-open this. 

The Batch Upload doesn't work and the location of those lines is back between 204-212.

https://github.com/google/google-api-php-client/blob/master/src/Google/Service/Resource.php#L210
